server/src/main.rs

```diff
@@ -0,0 +1,201 @@
+mod jump_definition;
+pub mod slicec_ext;
+
+#[tokio::main]
+async fn main() {
+    let stdin = tokio::io::stdin();
+    let stdout = tokio::io::stdout();
+
+    let (service, socket) = LspService::new(|client| Backend {
+        root_uri: Arc::new(Mutex::new(None)),
+        client,
+        compilation_state: Arc::new(Mutex::new(CompilationState::create())),
+        compilation_options: Arc::new(Mutex::new(SliceOptions::default())),
+    });
+
+    Server::new(stdin, stdout, socket).serve(service).await;
+}
+
+#[derive(Debug)]
+struct Backend {
+    client: Client,
+    root_uri: Arc<Mutex<Option<Url>>>,
+    compilation_state: Arc<Mutex<CompilationState>>,
+    compilation_options: Arc<Mutex<SliceOptions>>,
+}
+
+#[tower_lsp::async_trait]
+impl LanguageServer for Backend {
+    async fn initialize(
+        &self,
+        params: InitializeParams,
+    ) -> tower_lsp::jsonrpc::Result<InitializeResult> {
+        *self.root_uri.lock().await = params.root_uri;
+        Ok(InitializeResult {
+            server_info: None,
+            capabilities: ServerCapabilities {
+                text_document_sync: Some(TextDocumentSyncCapability::Options(
+                    TextDocumentSyncOptions {
+                        open_close: Some(true),
+                        change: Some(TextDocumentSyncKind::FULL), // or TextDocumentSyncKind::INCREMENTAL
+                        save: Some(TextDocumentSyncSaveOptions::SaveOptions(SaveOptions {
+                            include_text: Some(true),
+                        })),
+                        ..Default::default()
+                    },
+                )),
+                definition_provider: Some(OneOf::Left(true)),
+                hover_provider: Some(HoverProviderCapability::Simple(true)),
+                ..ServerCapabilities::default()
+            },
+        })
+    }
+
+    async fn initialized(&self, _: InitializedParams) {
+        // Compile all files to get the initial compilation state
+        let (initial_state, options) = self.compile_slice_files().await;
+        let mut compilation_state_guard = self.compilation_state.lock().await;
+        *compilation_state_guard = initial_state;
+        let mut compilation_options_guard = self.compilation_options.lock().await;
+        *compilation_options_guard = options;
+
+        self.client
+            .log_message(MessageType::INFO, "Slice language server initialized!")
+            .await;
+    }
+
+    async fn shutdown(&self) -> tower_lsp::jsonrpc::Result<()> {
+        Ok(())
+    }
+
+    async fn goto_definition(
+        &self,
+        params: GotoDefinitionParams,
+    ) -> tower_lsp::jsonrpc::Result<Option<GotoDefinitionResponse>> {
+        let uri = params.text_document_position_params.text_document.uri;
+        let position = params.text_document_position_params.position;
+        let state = self.compilation_state.lock().await;
+
+        if let Some(location) = get_definition_span(&state, uri, position) {
```

Comment:
This function also always returns `Ok` and has the `None` case map to `None`, so you could similarly refactor it. Just helps reduce redundancy is all.